oslo/torch/distributed/parallel_context.py

```diff
@@ -157,6 +157,22 @@ class ParallelContext(object):
         >>> parallel_context.get_prev_global_rank(ParallelMode.DATA)
     """
 
+    _instance = None
```

Comment:
Why did you make this class as singleton?
This is not our initial intention.

Reply:
Let's guess two models like BERT/GPT exist.
If you want to train BERT with TP, GPT with PP, how can you train the two models in one script?
Yes, It's not a usual situation. but if there's no reason, I don't think we must make it singleton.

Reply:
It seems that the current implementation of the parallel context in `oslo` only allows for creating a maximum of one instance. This is because `init_process_group` is called every time an instance is created, resulting in a RuntimeError: Trying to initialize the process group after it has already been initialized. It might be beneficial to make this explicit as a singleton, similar to ColossalAI's parallel context (https://github.com/hpcaitech/ColossalAI/blob/cd142fbefa964d62048a9bafb180322369ab89f8/colossalai/context/parallel_context.py#L24). The reason I implemented it this way is that the newly introduced DistributedTensor refers to the global context as distpec when the context is not explicitly specified. 

My suggestion is as follows (although I'm not sure if this is the right direction for discussion in this PR):

1. Make the current implementation a singleton, as it only allows for a maximum of one context.
2. If there is already a process group, do not create a new one.
3. If there is a process group, destroy it and create a new one.

Please let me know if my understanding is correct, and I will incorporate feedback accordingly :)

Reply:
Can we talk about this today? @yhna940